apps/server/src/modules/user-migration/service/user-migration.service.spec.ts

```diff
@@ -292,4 +330,159 @@ describe('UserMigrationService', () => {
+
+				await service.migrateUser('userId', 'externalUserTargetId', targetSystemId);
+
+				expect(userService.findById).toHaveBeenCalledWith('userId');
+				expect(userService.save).toHaveBeenCalledWith(migratedUserDO);
+				expect(accountService.findByUserIdOrFail).toHaveBeenCalledWith('userId');
+				expect(accountService.save).toHaveBeenCalledWith(migratedAccount);
+			});
+
+			it('should do migration of user', async () => {
+				const { migratedUserDO, notMigratedUser, accountDto, targetSystemId } = setupMigrationData();
+				userService.findById.mockResolvedValue(notMigratedUser);
+				accountService.findByUserIdOrFail.mockResolvedValue(accountDto);
+
+				await service.migrateUser('userId', 'externalUserTargetId', targetSystemId);
+
+				expect(userService.save).toHaveBeenCalledWith(migratedUserDO);
+			});
+
+			it('should do migration of account', async () => {
+				const { notMigratedUser, accountDto, migratedAccount, targetSystemId } = setupMigrationData();
+				userService.findById.mockResolvedValue(notMigratedUser);
+				accountService.findByUserIdOrFail.mockResolvedValue(accountDto);
+
+				await service.migrateUser('userId', 'externalUserTargetId', targetSystemId);
+
+				expect(accountService.save).toHaveBeenCalledWith(migratedAccount);
+			});
+		});
+
+		describe('when migration step failed', () => {
+			it('should throw Error', async () => {
+				const targetSystemId = new ObjectId().toHexString();
+				userService.findById.mockRejectedValue(new NotFoundException('Could not find User'));
+
+				await expect(service.migrateUser('userId', 'externalUserTargetId', targetSystemId)).rejects.toThrow(
+					new NotFoundException('Could not find User')
+				);
+			});
+
+			it('should log error and message', async () => {
+				const { migratedUserDO, accountDto, targetSystemId } = setupMigrationData();
+				const error = new NotFoundException('Test Error');
+				userService.findById.mockResolvedValue(migratedUserDO);
+				accountService.findByUserIdOrFail.mockResolvedValue(accountDto);
+				accountService.save.mockRejectedValueOnce(error);
+
+				await service.migrateUser('userId', 'externalUserTargetId', targetSystemId);
+
+				expect(logger.log).toHaveBeenCalledWith(error);
+				expect(logger.log).toHaveBeenCalledTimes(2);
+			});
+
+			it('should do a rollback of migration', async () => {
+				const { notMigratedUser, accountDto, targetSystemId } = setupMigrationData();
+				const error = new NotFoundException('Test Error');
+				userService.findById.mockResolvedValue(notMigratedUser);
+				accountService.findByUserIdOrFail.mockResolvedValue(accountDto);
+				accountService.save.mockRejectedValueOnce(error);
+
+				await service.migrateUser('userId', 'externalUserTargetId', targetSystemId);
+
+				expect(userService.save).toHaveBeenCalledWith(notMigratedUser);
+				expect(accountService.save).toHaveBeenCalledWith(accountDto);
+			});
+
+			it('should return to dashboard', async () => {
```

Comment:
where is the dashboard url coming from?

The server should not be client specific. Depending on what client is used, /dashboard is a route that may or may not exist, and wether or not it does, the server has no reason at all to know.

If this is an url provided by an external service its fine to blindly pass a redirect url. But the formulation of this test indicates that this is actual business logic that defines its own redirect url, and that looks like a breach of server responsability

Reply:
see comment in user-migration.service.ts